File: JRPG-character/src/ru/morecode/JRPG/character/Effect.java
package ru.morecode.JRPG.character;

public class Effect {
    String type;
    int tier;
    public Effect(String type, int tier){
        this.type = type;
        this.tier = tier;
    }
    public void doEffect(Character target){
        switch (type){
            case "лечение":{
                Heal(target,tier);
                break;
            }
        }
    }
    public void Heal(Character target, int tier){
        int heal = 10 * tier;
        if ( target.maxHealth - target.health <= heal) {
            target.health = target.maxHealth;
        } else {
            target.health += heal;
        }
    }
}

File: QT_JRPG/JRPG/effect.cpp
#include "effect.h"

Effect::Effect(int id, QString type, int tier)
{
    this->id;
    this->type = type;
    this->tier = tier;
}

void Effect::doEffect(Character* user, Character* target){
    if(target == nullptr){
    switch (toIntType(type)){
    case HEAL:{
        Heal(user,tier);
        break;
    }
    }
    }
}
void Effect::Heal(Character* target, int tier){
    int vitality = target->getVitality();
    int wisdom = target->getWisdom();

    int heal = (vitality + wisdom / 2) * tier;

    int health = target->getHealth();
    int maxHealth = target->getMaxHealth();

    if ( maxHealth - health <= heal) {
        target->setHealth(maxHealth);
    } else {
        target->setHealth(health + heal);
    }
}

int Effect::toIntType(QString type)
{
    if(type == "лечение"){
        return HEAL;
    }
    return -1;
}
